src/ImmersedBoundaries/grid_fitted_immersed_boundaries.jl

```diff
@@ -67,21 +68,37 @@ Return a grid with `GridFittedBottom` immersed boundary.
 Computes ib.bottom_height and wraps in an array.
 """
 function ImmersedBoundaryGrid(grid, ib::AbstractGridFittedBottom)
-    bottom_field = Field{Center, Center, Nothing}(grid)
+    bottom_field = Field((Center, Center, Nothing), grid)
     set!(bottom_field, ib.bottom_height)
     fill_halo_regions!(bottom_field)
     offset_bottom_array = dropdims(bottom_field.data, dims=3)
 
-    # TODO: maybe clean this up
-    IB = typeof(ib).name.wrapper
-    new_ib = IB(offset_bottom_array)
+    new_ib = getname(ib)(offset_bottom_array)
 
     return ImmersedBoundaryGrid(grid, new_ib)
 end
 
 function ImmersedBoundaryGrid(grid, ib::AbstractGridFittedBottom{<:OffsetArray})
     TX, TY, TZ = topology(grid)
-    # TODO: check size
+    N = size(grid)
+    H = halo_size(grid)
+
+    field_size = N[1:2] .+ 2 .* H[1:2]
+
+    # If using `with_halo` in combination with a `DistributedArch`, we need to make 
+    # sure that halos are filled correctly for the `ib.bottom_height`
+    if any(size(ib.bottom_height) .!= field_size)
+        bottom_field = Field((Center, Center, Nothing), grid)
+        cpu_array    = arch_array(CPU(), ib.bottom_height)[1:N[1], 1:N[2]] 
+        set!(bottom_field, cpu_array)
+        fill_halo_regions!(bottom_field)
+        offset_bottom_array = dropdims(bottom_field.data, dims=3)
+    
+        new_ib = getname(ib)(offset_bottom_array)
+    
+        return ImmersedBoundaryGrid(grid, new_ib)
+    end
+
```

Comment:
Should we write a new function for a `grid` on `DistributedArch` instead?

Reply:
I'm not sure why that `if` statement will always only be triggered for distributed arch

Reply:
So I think it's better to be explicit. We could also write

```julia
if architecture(grid) isa DistributedArch
```

for example

Reply:
this might be helpful also as a safety net for non distributed architectures when the OffsetArray passed does not have correct halos

Reply:
I think we want to _avoid_ that "silent safety" because `ImmersedBoundaryGrid(grid, ib::AbstractGridFittedBottom{<:OffsetArray})` is an internal function, not a user facing function. Such a safety net creates ambiguity about the user interface; users can make a mistake but not receive an error. We want to avoid that.

Reply:
In other words I think it is preferable here to receive an error if there is some mix-up about the grid halos... every experiment should use just one grid.

Reply:
Using `if architecture(grid) isa DistributedArch` will help preserve that error.

Reply:
So I think we should leave the check but just spit an error, I moved the resizing in the Distributed module